### DISCUSSION TITLE
A new type of primitive I've made for CSharpCAD, that might be of interest to JSCAD.

### DISCUSSION THREAD

**Discussion by briansturgill** [jscad/OpenJSCAD.org#1104] [General]
So these two primitives are what I'm calling a "negative geometry object"... basically it's an object that is created only so that it can be subtracted from another object (yielding a presumably useful object).

(BTW, if you've ever seen a package contain "negative geometry objects" before, I'd love to know about it!)

I created Cutter2D and Cutter 3D. Their function is to remove a "pie slice" from circular/elliptical objects.

One of the few things I like better about OpenSCAD over JSCAD is that it's API is more simple.
Circle for example has those annoying startAngle and endAngle options.
Within JSCAD, the complexity of having Semi and Full circles in the same API has caused problems.

As no one is using CSharpCAD, I can make reasonable changes in the API... and this allows me to simplify all the circular/elliptical APIs, yet still be able for end users to make semi-objects easily.

JSCAD may want to adopt Cutter2D/3D, even if the APIs are not changed as it works with GeodesicSphere, Ellipsoid and Sphere, which don't have semi-versioning in their API. And also on any round compound objects.

Example:
```
var g = Subtract(Cylinder(height: 10, radius: 5), Cutter3D(radius: 5, angle: 90));
```

Here's my C# implementation.  Remember to translate to radians! (CSharpCAD uses degrees.)

```C#
namespace CSharpCAD;

public static partial class CSCAD
{
    /// <summary>Construct a "cutter", a negative geometry object, suitable to cut a "pie slice" out of a circlular 2d object of radius r.</summary>
    /// <remarks>
    /// This object would normally be used as a second argument to "Subtract".
    /// To use with an elliptical object, specify radius as the largest radius of the elliptical.
    /// </remarks>
    /// <param name="radius">Radius that matches the circle you want to cut.</param>
    /// <param name="angle" default="90">Angle formed by the cut (in degrees).</param>
    /// <param name="center" default="(0,0)">Center that coincides with the center of the circular target.</param>
    /// <example>
    /// // Cuts the fourth quadrant of the circle away.
    /// var g = Subtract(Circle(radius: 5), Rotate((0, 0, 90), Cutter2D(radius: 6, angle: 90)));
    /// </example>
    /// <group>2D Primitives</group>
    public static Geom2 Cutter2D(double radius = 1, double angle = 90, Vec2? center = null)
    {
        if (radius <= 0.0) throw new ArgumentException("Radius value must be postive.");
        if (LessThanish(angle, 1)) throw new ArgumentException("Angle must be greater than one.");
        if (GreaterThanish(angle, 360)) throw new ArgumentException("Angle must be less than a full circle.");
        var _center = center ?? new Vec2(0, 0);

        radius = radius * 2 + 1; // Using +1 to make sure we are bigger than the diameter.

        var points = new List<Vec2>(5); // 3 90 degree points, and possibly an end point and a center point
        void addPoint(double x, double y) => points.Add(_center.Add(new Vec2(x, y)));

        if (!Equalish(angle, 180.0)) points.Add(_center); // Can't add addPoint, also, if 180ish then colinear with start and end.

        points.Add(new Vec2(radius * Cos(0), radius * Sin(0)));
        for (var pointAngle = 90.0 ; LessThanish(pointAngle, angle); pointAngle += 90.0)
        {
            addPoint(radius * Cos(pointAngle), radius * Sin(pointAngle));
        }
        addPoint(radius * Cos(angle), radius * Sin(angle));

        var cutter2D = new Geom2(points);
        return cutter2D;
    }

    /// <summary>Construct a "cutter", a negative geometry object, suitable to cut a "pie slice" out of a circlular 3D object of radius r.</summary>
    /// <remarks>
    /// This object would normally be used as a second argument to "Subtract".
    /// To use with an elliptical object, specify radius as the largest radius of the elliptical.
    /// </remarks>
    /// <param name="radius">Radius that matches the circular 3D object you want to cut.</param>
    /// <param name="height" default="radius*2+2">Height that matches the cylinder or other circular 3D object you want to cut.</param>
    /// <param name="angle" default="90">Angle formed by the cut (in degrees).</param>
    /// <param name="center" default="(0,0,0)">Center that coincides with the center of the circular target.</param>
    /// <example>
    /// var g = Subtract(Cylinder(height: 10, radius: 5), Cutter3D(radius: 5, angle: 90));
    /// </example>
    /// <group>3D Primitives</group>
    public static Geom3 Cutter3D(double radius = 1, double? height = null, double angle = 90, Vec3? center = null)
    {
        if (radius <= 0.0) throw new ArgumentException("Radius value must be postive.");
        if (LessThanish(angle, 1)) throw new ArgumentException("Angle must be greater than one.");
        if (GreaterThanish(angle, 360)) throw new ArgumentException("Angle must be less than a full circle.");
        var _center = center ?? new Vec3(0, 0, 0);
        var _height = height ?? radius * 2;

        _height = _height + 2; // Make sure we cut everything.

        var cutter2D = Cutter2D(radius, angle);
        var cutter3D = (Geom3)Translate(new Vec3(_center.X, _center.Y, _center.Z - (_height / 2.0)), ExtrudeLinear(cutter2D, height: _height));
        return cutter3D;
    }
}
```

**Comment by hrgdavor** [comment]
```
var g = Subtract(Circle(radius: 5), Rotate((0, 0, 90), Cutter2D(radius: 6, angle: 90)));
```
I am curious
 - does radius have to be the same, if it can differ, how it affects end-result?
 - can there be arbitrary rotation, or it needs to be only on one axis ?

**Reply by briansturgill** [reply]
The radius can be any number >= radius of the circle you're cutting.
Choosing the largest radius from an elliptical guarantees a clean cut.

Think of the "Cutter" as a cookie cutter for pie wedges (where pie wedges can go up to 360-EPS degrees).
It has an odd shape (a polygon of 3-6 vertices, extruded for 3D) because I wanted it to have as few points as possible for efficiency.

You can think of it as looking like a pie slice (in 2D and 3D)... however for efficiency I allow at most 4 "segments".
For fun, you can look at SVG output of a 300, 270, 180 and 90 degree "cutters". The 3D look just like them, but extruded to height.
Oh... have a 45 degree one up on my screen:
![image](https://user-images.githubusercontent.com/7350143/172956241-7756cd73-f76b-47a2-9621-ecb3825c7385.png)
At the left is the circle, the cutter, then the subtraction.
A 180 is the simplest case. It is only 3 points. start, end and one at 90deg. (Most have a point to center, but that point is colinear
with start and end and thus dropped from the polygon.

Here is 300 degrees:
![image](https://user-images.githubusercontent.com/7350143/172956904-cd0b6a4c-692d-456f-83ab-02241fd3b038.png)


Following the primitives from JSCAD, I'm using the x/y plane as the circular and the height is z.
Think Cylinder and Circle for the correct mind model.
This was intended as a replacement for the startAngle/endAngle options in JSCAD, with the added benefit
that it works on GeodesicSphere, Sphere, Ellipsoid and any compound circular object.

Say you wanted to 3D print a snow man that sits on a corner of a house. Make the model, then subtract a Cutter3d using the largest radius of the snowman (guess large if you need to) and the height of the snowman (again you can guess large, won't hurt anything) and 90 degrees (might want to use 91) to define the corner.
Anyway, "subtract" the Cutter3d from the snowman object and you have the desired result.

Rotation will always be around the Z for things similar to circles and cylinders.
Again, this is exactly the same as when using startAngle/endAngle in JSCAD.
That said, on a round in 3D figure, like a Sphere, you really could rotate on any one axis.
That is likely more interesting on an ellipsoid than on spheres.

**Reply by briansturgill** [reply]
Oh, to complete the equivalence.
angle on a cutter is endAngle-startAngle in JSCAD.
If startAngle is not zero in JSCAD, then rotate the Z-axis by startAngle.

**Reply by hrgdavor** [reply]
start and end angles are procedural options (directly calculated and fast).

I may be misunderstanding ...
Your approach looks to me will be using booelans to get the same effect which would be much slower (worse with each segments increase).

**Reply by briansturgill** [reply]
It is true I'm removing some functionality and replacing it with somewhat slower... but at least in the case of JSCAD, I'm not saying you need to do that (though I'd run some timing tests... that functionality, rarely used, isn't free).
What I'm saying is those two primitives I made will give you startAngle/endAngle functionality on both kinds of spheres and on ellipsoids... plus it can work on compound objects. It is not either or... you can keep startAngle/endAngle AND use those new primitives when useful. I think for CSharpCAD that the simplification of the APIs make it worth the slowness, but I'll try some tests first. I don't think the subtracts will be that onerous, especially as they are rarely used anyway.

**Reply by briansturgill** [reply]
If you have to keep startAngle/endAngle, consider making the more commonly used functions (circle, ellipse and cylinder e.g.) and making a Circle and SemiCircle, etc., where Circle is ALWAYS a full circle and SemiCiricle IS NEVER a full circle. It will be easier to maintain and things like Circle are heavily used. It would also be easier for beginners to understand the API.

**Reply by briansturgill** [reply]
@hrgdavor OK, so I ran timing tests (100 runs). To use the JSCAD API (i.e. startAngle/endAngle) is about 10 times faster than doing it with Cutter... but it only takes 1ms per creation. How many Circles using startAngle/endAngle do you think you'll make in a real project?
I guess my point is that I'm all into getting things to run fast, but it's not the only thing. And something that reasonably would add at most a few milliseconds to your overall create time, just isn't important.  Understand that either way the Semi-Circle is created, the resulting Geom2 is essentially the same.
When I talked about my recent translation of a OpenSCAD project to CSharpCAD. The difference was 2 seconds vs 24 seconds. Now that is a big deal when you're working on a design.

**Reply by briansturgill** [reply]
@hrgdavor I just added 3D testing with CylinderElliptic. It takes about 2ms instead of 1 ms in the 2D case.jjj
Sadly startAngle/endAngle is taking 1ms (vs 0.1ms in 2D) so, the Cutter approach is clearly acceptable in real world usage.

**Reply by hrgdavor** [reply]
Booleans are great for doing general stuff, but come with performance cost that is exponential to number of segments. I actually prefer procedural solutions for specific common cases which are limited in scope but get more dramatically faster each segments increment. It is not unusual for me to render with 128 segments to get smoother shapes for 3d print.

It is not "one or the other" for me, I would use both depending on the use-case.

**Reply by briansturgill** [reply]
Proper booleans are not exponential to number of segments. They should be roughly n log(n)... JSCAD's may have appeared exponential, but replacing Math.hypot will fix much of that issue. But the number of segments (really vertices) does matter, thus my working to make Cutter uses as few vertices as possible.

**Reply by hrgdavor** [reply]
the more segments a cylinder has, there are more planes that intersect, it gets much worse with sphere ofc.

take this simple example
```
subtract(
    sphere({}),
    translate([0,0.5,0.5],sphere({segments}))
  )
```
first sphere is left with 32 segments
then for sphere 2:
 - 32 segments - 80ms
 - 64 segments - 380ms
 - 128 segments - 3400ms
I doubt hypot is causing this

and to repeat if is not clear, I am not against utility like cutter. It is my personal preference to do more things procedurally, because booleans get bogged down quickly.

**Reply by briansturgill** [reply]
Well the good thing about cutter is that it doesn't grow. You have at most 6 vertices.

I'll run that same sphere test under CSCAD to check if it's an implementation issue, or an algorithm issue.
If it's algorithm we should look for a new one, as better ones supposedly exist.

**Reply by briansturgill** [reply]
Starting
Sphere 32 loops: 1: 83ms
Starting
Sphere 64 loops: 1: 318ms
Starting
Sphere 128 loops: 1: 2782ms

Perhaps the algorithm does need looked at (again, I've done that at least 5 times).
It occurs to me that possibly it's garbage collection related. Generally GC gets ignored in academic algorithm analysis.
I say this because the change in time between 32 and 64 is reasonable.

**Reply by briansturgill** [reply]
@hrgdavor Well, here's an argument for leaving startAngle/endAngle in CSharpCAD:
Starting
Subtract(Sphere(radius: 10, segments: 128), Cutter3D(radius: 10, angle: 90)); loops: 1: 2444ms
For Spheres, it's your only option, but clearly making Cutter's output small isn't the overriding concern.
Make's me believe even more that the problem is memory management, not algorithm.

**Reply by briansturgill** [reply]
@hrgdavor No, no, no... I can't believe I didn't spot this immediately, the problem isn't with the algorithm... it's the object chosen.
Sphere by definition is quadrupling the points at each doubling.
Vertex count:
Sphere(32) 1984
Sphere(64) 8064
Sphere(128) 32512

**Reply by hrgdavor** [reply]
I doubt that gc is culprit in 64 to 128, but if nothing else, it shows that current boolean speed is not logarithmic compared to point count :)

**Reply by briansturgill** [reply]
@hrgdavor Yes, but you were increasing point count exponentially when you doubled segments. (I.E. in a sphere, doubling segments, quadruples vertices. So, of course, time went up exponentially.)
Here is a paper by Joseph O'Rourke (a well known academic in the area and author of Computation Geometry in C.)
He is claiming an O(n+m) algorithm for the intersection of 2 polygons (n and m the number of vertices in the two polygons).
Intersection algorithms can do union and difference with minor modifications.
https://www.cs.jhu.edu/~misha/Spring16/ORourke82.pdf
He talks about the Shamos O(n*log(n)) which is the algorithm I was referring to above.

If you really believe JSCAD booleans are O(n^2) or worse, then something is very wrong.
I suspected this, but currently believe that things like Math.Hypot was making the "constant" high.
O(n) means linear, but really is a a short hand for C*n+K where C and K are "constants"... but if high, can dominate the algorithm.
(C would refer to operations conducted on each n, K would be things like setup/teardown time.)

**Comment by briansturgill** [comment]
@hrgdavor Putting this in a new thread.
Here is an intensive test of the Boolean subtract. It will largely hold true for Union and Intersect as well.
I think it shows a real world implementation of an O(n*log(n)) algorithm. As things get truly large, a combination of garbage
 collection and cache misses caused by the shear volume of what is going on makes things get slightly worse over time.
 
```
Starting
Subtract(CylinderElliptic(startRadius: (10, 10), endRadius: (10, 10), segments: 32), Cutter3D(radius: 10, angle: 90)); loops: 1: 97ms
Starting
Subtract(CylinderElliptic(startRadius: (10, 10), endRadius: (10, 10), segments: 64), Cutter3D(radius: 10, angle: 90)); loops: 1: 1ms
Starting
Subtract(CylinderElliptic(startRadius: (10, 10), endRadius: (10, 10), segments: 128), Cutter3D(radius: 10, angle: 90)); loops: 1: 4ms
Starting
Subtract(CylinderElliptic(startRadius: (10, 10), endRadius: (10, 10), segments: 256), Cutter3D(radius: 10, angle: 90)); loops: 1: 10ms
Starting
Subtract(CylinderElliptic(startRadius: (10, 10), endRadius: (10, 10), segments: 512), Cutter3D(radius: 10, angle: 90)); loops: 1: 34ms
Starting
Subtract(CylinderElliptic(startRadius: (10, 10), endRadius: (10, 10), segments: 1024), Cutter3D(radius: 10, angle: 90)); loops: 1: 107ms
Starting
Subtract(CylinderElliptic(startRadius: (10, 10), endRadius: (10, 10), segments: 2048), Cutter3D(radius: 10, angle: 90)); loops: 1: 305ms
Starting
Subtract(CylinderElliptic(startRadius: (10, 10), endRadius: (10, 10), segments: 4096), Cutter3D(radius: 10, angle: 90)); loops: 1: 914ms
Starting
Subtract(CylinderElliptic(startRadius: (10, 10), endRadius: (10, 10), segments: 8192), Cutter3D(radius: 10, angle: 90)); loops: 1: 2424ms
Starting
Subtract(CylinderElliptic(startRadius: (10, 10), endRadius: (10, 10), segments: 16384), Cutter3D(radius: 10, angle: 90)); loops: 1: 6427ms
Starting
Subtract(CylinderElliptic(startRadius: (10, 10), endRadius: (10, 10), segments: 32768), Cutter3D(radius: 10, angle: 90)); loops: 1: 22192ms
Starting
Subtract(CylinderElliptic(startRadius: (10, 10), endRadius: (10, 10), segments: 65536), Cutter3D(radius: 10, angle: 90)); loops: 1: 84821ms
Starting
Subtract(CylinderElliptic(startRadius: (10, 10), endRadius: (10, 10), segments: 131072), Cutter3D(radius: 10, angle: 90)); loops: 1: 354742ms
```
As expected Cyllinder Elliptic grows linearly with the number of segments:
CylinderElliptic(32) 320
CylinderElliptic(64) 640
CylinderElliptic(128) 1280
CylinderElliptic(256) 2560
CylinderElliptic(512) 5120
CylinderElliptic(1024) 10240
CylinderElliptic(2048) 20480
CylinderElliptic(4096) 40960
CylinderElliptic(8192) 81920
CylinderElliptic(16384) 163840
CylinderElliptic(32768) 327680
CylinderElliptic(65536) 655360
CylinderElliptic(131072) 1310720

**Reply by hrgdavor** [reply]
@briansturgill thanks for the extensive tests. It gives a better insight into how booleans scale :) 
Also thanks for the link to the paper. My knowledge on computational geometry is bit thin :)

If you have more to share, please do :) :) 

My claims were based on experience and was guesswork (it is a big nono to make assumptions about performance) :)

hope I did not waste too much of your time :)

**Reply by briansturgill** [reply]
No waste of my time at all, I've been meaning to do a more thorough timing test of the booleans anyway. (Nothing like actually checking to see if you understand things correctly.) As to the startAngle/endAngle, I'm still not fully decided what to do. I have some more timing tests to do. I want to see if the presence of the code and the sharing of code between like circle/ellipse have significant effects. I'm sure it does somewhat in JSCAD because of the work you have to do with handling options, though I'm not sure just how significant that is. I also noticed an anomaly where circle + extrude was faster (a lot) than cylinderElliptic. I keep thinking I'm missing something, but at least the output of the two tests matched in prusaslicer.  Anyway, as always, if I find anything significant, I'll let JSCAD people know!

**Reply by hrgdavor** [reply]
Cool, I really like the insights you bring here. Can't wait to free-up some time to do more jscad stuff. Currently all I can manage is comment a bit here and there.

**Comment by receter** [comment]
I have made a similar function some time ago. It is called `cuttingMachine` and it looks like this:

<img width="1236" height="210" alt="grafik" src="https://github.com/user-attachments/assets/a38c0203-be20-4c2a-acaf-ae15db179d9b" />

You can also try it here https://stl.parts/shape/cuboid/

Furthermore the processing steps have a "drillingMachine". The function takes an shape, attaches it to the top of an object with measureBoundingBox and uses hull to drill into a part. In the processing steps the function is always a cylinder.
